src/Components/Endpoints/src/Discovery/ComponentApplicationBuilder.cs

```diff
@@ -9,13 +9,25 @@ namespace Microsoft.AspNetCore.Components;
 public class ComponentApplicationBuilder
```

Comment:
I don't really have any intuition for how much of these APIs for discovery and combining/excluding libraries is strictly needed or is for edge cases.

 * To what extent is this the same as patterns we have elsewhere?
 * Is the terminology of "library" something invented here, or are you taking that from something else in ASP.NET?
 * If we were determined to strip this down to the absolute minimum needed to make something shippable in 8.0, what would be left?
 * Why does it have to keep track of components anyway? I understand it has to track pages since it's the thing providing the list of endpoints.

Reply:
We need to support things like adding "areas" dynamically, for example if we have a Blazor based version of Identity UI, and we need to have ways of removing components/pages from a given app. https://github.com/dotnet/aspnetcore/issues/46980#issuecomment-1460509030

We track the components because we use that information to determine the component render mode (for now) but it is also going to include more data in the future, like parameters that require model binding, etc. that is coming from https://github.com/dotnet/aspnetcore/pull/48284

Reply:
> To what extent is this the same as patterns we have elsewhere?
This is a "slimmed down" design compared to MVC's application parts.

> Is the terminology of "library" something invented here, or are you taking that from something else in ASP.NET?

We used the `Part` suffix in MVC, but I did not want to get that route, so I settled for Library as in `Razor Class Library`. In the end, the library is the collection of data that we have constructed either through source generation or by reflecting over the assembly.